sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -495,22 +495,45 @@ template <typename T, uint8_t NElts = 1,
           cache_hint L1H = cache_hint::none, cache_hint L3H = cache_hint::none>
 __ESIMD_API __ESIMD_NS::simd<T, NElts>
 lsc_block_load(const T *p, __ESIMD_NS::simd_mask<1> pred = 1) {
-  detail::check_lsc_vector_size<NElts>();
   detail::check_lsc_data_size<T, DS>();
   detail::check_lsc_cache_hint<detail::lsc_action::load, L1H, L3H>();
   constexpr uint16_t _AddressScale = 1;
   constexpr int _ImmOffset = 0;
   constexpr lsc_data_size _DS = detail::finalize_data_size<T, DS>();
-  static_assert(_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64,
-                "Transposed load is supported only for data size u32 or u64");
-  constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
   constexpr detail::lsc_data_order _Transposed =
       detail::lsc_data_order::transpose;
   constexpr int N = 1;
   __ESIMD_NS::simd<uintptr_t, N> addrs = reinterpret_cast<uintptr_t>(p);
-  return __esimd_lsc_load_stateless<T, L1H, L3H, _AddressScale, _ImmOffset, _DS,
-                                    _VS, _Transposed, N>(pred.data(),
-                                                         addrs.data());
+  if constexpr (_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64) {
+    detail::check_lsc_vector_size<NElts>();
+    constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
+    return __esimd_lsc_load_stateless<T, L1H, L3H, _AddressScale, _ImmOffset,
+                                      _DS, _VS, _Transposed, N>(pred.data(),
+                                                                addrs.data());
+  } else if constexpr (_DS == lsc_data_size::u16) {
+    static_assert(NElts % 2 == 0,
+                  "Number of elements is not supported by Transposed load");
+    detail::check_lsc_vector_size<NElts / 2>();
+    constexpr detail::lsc_vector_size _VS =
+        detail::to_lsc_vector_size<NElts / 2>();
+
+    __ESIMD_NS::simd<uint32_t, NElts / 2> result =
+        __esimd_lsc_load_stateless<uint32_t, L1H, L3H, _AddressScale,
+                                   _ImmOffset, lsc_data_size::u32, _VS,
+                                   _Transposed, N>(pred.data(), addrs.data());
+    return result.template bit_cast_view<T>();
+  } else if constexpr (_DS == lsc_data_size::u8) {
+    static_assert(NElts % 4 == 0,
+                  "Number of elements is not supported by Transposed load");
+    detail::check_lsc_vector_size<NElts / 4>();
+    constexpr detail::lsc_vector_size _VS =
+        detail::to_lsc_vector_size<NElts / 4>();
+    __ESIMD_NS::simd<uint32_t, NElts / 4> result =
+        __esimd_lsc_load_stateless<uint32_t, L1H, L3H, _AddressScale,
+                                   _ImmOffset, lsc_data_size::u32, _VS,
+                                   _Transposed, N>(pred.data(), addrs.data());
+    return result.template bit_cast_view<T>();
+  }
```

Comment:
There are 2 identical blocks of code. The only difference there is 2 vs 4 factor.
Please combine them to something like this:
```suggestion
  } else if constexpr (_DS == lsc_data_size::u16 || _DS == lsc_data_size::u8) {
    constexpr NElemsInDword = _DS == lsc_data_size::u16 ? 2 : 4;
    static_assert(NElts % NElemsInDword == 0,
                  "Number of elements is not supported by Transposed load");
    detail::check_lsc_vector_size<NElts / NElemsInDword>();
    constexpr detail::lsc_vector_size _VS =
        detail::to_lsc_vector_size<NElts / NElemsInDword>();

    __ESIMD_NS::simd<uint32_t, NElts / NElemsInDword> result =
        __esimd_lsc_load_stateless<uint32_t, L1H, L3H, _AddressScale,
                                   _ImmOffset, lsc_data_size::u32, _VS,
                                   _Transposed, N>(pred.data(), addrs.data());
    return result.template bit_cast_view<T>();
  }
```